crates/shared/src/sources/balancer_v2/swap.rs

```diff
@@ -115,27 +144,39 @@ impl BaselineSolvable for WeightedPoolRef<'_> {
         out_reserves.common.downscale_down(out_amount)
     }
 
-    fn get_amount_in(&self, in_token: H160, (out_amount, out_token): (U256, H160)) -> Option<U256> {
-        // Note that the output of this function does not depend on the pool
-        // specialization. All contract branches compute this amount with:
-        // https://github.com/balancer-labs/balancer-v2-monorepo/blob/6c9e24e22d0c46cca6dd15861d3d33da61a60b98/pkg/core/contracts/pools/BaseMinimalSwapInfoPool.sol#L75-L88
+    fn get_amount_in(
+        &self,
+        in_token: H160,
+        (exact_out_amount, out_token): (U256, H160),
+    ) -> Option<U256> {
+        // Balancer V2 weighted pools are "unstable", where if you compute an
+        // input amount large enough to buy X tokens, selling the computed
+        // amount over the same pool in the exact same state will yield X-𝛿
+        // tokens. To work around this, for each hop, we try to converge to some
+        // sell amount >= the required buy amount.
+
+        let in_amount = self.get_amount_in_exact(in_token, (exact_out_amount, out_token))?;
+        let out_amount = self.get_amount_out(out_token, (in_amount, in_token))?;
+        if out_amount >= exact_out_amount {
+            return Some(in_amount);
+        }
+
+        // If the computed output amount is not enough; we bump the sell amount
+        // a bit. We start with 1e-12 tokens (normalized to 18 decimals) and
+        // multiply the amount to bump by 10 for each iteration up to 1e-6.
```

Comment:
Why these values? Are they chosen somewhat arbitrarily? If I sell Gemini USD (2 decimals) would we start bumping by $1000?

Maybe it's safe to go from -18 to 0 instead (would still only be 18 iterations)?

Reply:
@fleupold - can you please re-review. I made non trivial refactors to the original PR to try and better address your comments.